src/request-pipeline/destination-request/index.js

```diff
@@ -33,41 +32,60 @@ export default class DestinationRequest extends EventEmitter {
         this._send();
     }
 
-    _send () {
+    _send (waitData) {
         connectionResetGuard(() => {
             var timeout = this.opts.isXhr ? DestinationRequest.XHR_TIMEOUT : DestinationRequest.TIMEOUT;
 
-            this.req = this.protocolInterface.request(this.opts);
+            this.req = this.protocolInterface.request(this.opts, res => {
+                if (waitData) {
+                    res.on('data', () => void 0);
+                    res.on('end', () => this._onResponse(res));
+                }
+            });
+
+            if (!waitData)
+                this.req.on('response', res => this._onResponse(res));
 
-            this.req.on('response', res => this._onResponse(res));
             this.req.on('error', err => this._onError(err));
             this.req.setTimeout(timeout, () => this._onTimeout());
-
             this.req.write(this.opts.body);
             this.req.end();
         });
     }
 
-    _onResponse (res) {
-        this.hasResponse = true;
+    _canTryToAuthorize (res) {
+        if (this.opts.credentials) {
+            var authInfo = getAuthInfo(res);
 
-        if (res.statusCode === 401)
-            this._sendWithCredentials(res);
+            // NOTE: If we get 401 status code after the credential sent, we should to stop trying to authenticate.
+            if (!authInfo.isChallengeMessage && this.credentialsSent)
+                return false;
 
-        else
+            return authInfo.canAuthorize;
+        }
+
+        return false;
+    }
+
+    _onResponse (res) {
+        if (res.statusCode === 401 && this._canTryToAuthorize(res))
+            this._resendWithCredentials(res);
+        else {
+            this.hasResponse = true;
             this.emit('response', res);
+        }
     }
 
-    async _sendWithCredentials (res) {
-        if (this.opts.credentials) {
-            if (OS.win)
-                await assignWindowsDomain(this.opts.credentials);
+    async _resendWithCredentials (res) {
+        addCredentials(this.opts.credentials, this.opts, res, this.protocolInterface);
+        this.credentialsSent = true;
 
-            //TODO !!!
-            requestWithAuth(this.opts, this.opts.credentials, [this.opts.body], response => {
-                this.emit('response', response);
-            }, this.isHttps, res);
-        }
+        var authInfo = getAuthInfo(res);
+
+        // NOTE: The NTLM authentication requires to use one socket for the "negotiate" and "authenticate" requests.
+        // So, before "authenticate" message sending we should to wait data form the "challenge" response. It will
+        // mean that the socket is free.
```

Comment:
```
        // NOTE: NTLM authentication requires using the same socket for the "negotiate" and "authenticate" requests.
        // So, before sending the "authenticate" message, we should wait for data from the "challenge" response. It will
        // mean that the socket is free.
```